File: advent2017/extras/python/days/day18.py
#!/usr/bin/env python3
"""AoC 2017 Day 18 fancy Python implementation (duet VM)."""

from __future__ import annotations

import argparse
import hashlib
import time
from collections import deque
from pathlib import Path

EXPECTED_SHA256 = "4052bdd33baaf7be897365aa3ad1cff5fae76ade4c474c9e5ebcdf5058ad368e"
EXPECTED_PART1 = "7071"
EXPECTED_PART2 = "8001"


def default_input_path() -> Path:
    cands = [
        Path("advent2017/Day18/d18_input.txt"),
        Path("Day18/d18_input.txt"),
        Path("../Day18/d18_input.txt"),
        Path("../../Day18/d18_input.txt"),
        Path(__file__).resolve().parents[2] / "Day18" / "d18_input.txt",
    ]
    for c in cands:
        if c.exists():
            return c.resolve()
    raise FileNotFoundError("input not found for day 18")


def sha256_file(path: Path) -> str:
    h = hashlib.sha256()
    with path.open("rb") as f:
        for chunk in iter(lambda: f.read(1 << 20), b""):
            h.update(chunk)
    return h.hexdigest()


def parse(raw: str) -> list[tuple[str, str, str | None]]:
    out: list[tuple[str, str, str | None]] = []
    for line in raw.splitlines():
        if not line.strip():
            continue
        p = line.split()
        out.append((p[0], p[1], p[2] if len(p) > 2 else None))
    return out


def val(tok: str, regs: dict[str, int]) -> int:
    if tok.lstrip("-").isdigit():
        return int(tok)
    return regs.get(tok, 0)


def solve_part1(prog: list[tuple[str, str, str | None]]) -> int:
    regs: dict[str, int] = {}
    ip = 0
    last = 0

    while 0 <= ip < len(prog):
        op, x, y = prog[ip]
        if op == "snd":
            last = val(x, regs)
        elif op == "set":
            regs[x] = val(y or "0", regs)
        elif op == "add":
            regs[x] = regs.get(x, 0) + val(y or "0", regs)
        elif op == "mul":
            regs[x] = regs.get(x, 0) * val(y or "0", regs)
        elif op == "mod":
            regs[x] = regs.get(x, 0) % val(y or "1", regs)
        elif op == "rcv":
            if val(x, regs) != 0:
                return last
        elif op == "jgz":
            if val(x, regs) > 0:
                ip += val(y or "0", regs)
                continue
        ip += 1

    raise RuntimeError("no recovery")


class Proc:
    def __init__(self, pid: int, prog: list[tuple[str, str, str | None]]) -> None:
        self.pid = pid
        self.prog = prog
        self.regs: dict[str, int] = {"p": pid}
        self.ip = 0
        self.inq: deque[int] = deque()
        self.send_count = 0
        self.waiting = False
        self.terminated = False

    def step(self, outq: deque[int]) -> bool:
        if not (0 <= self.ip < len(self.prog)):
            self.terminated = True
            self.waiting = False
            return False

        op, x, y = self.prog[self.ip]

        if op == "snd":
            outq.append(val(x, self.regs))
            self.send_count += 1
            self.ip += 1
            self.waiting = False
            return True
        if op == "set":
            self.regs[x] = val(y or "0", self.regs)
            self.ip += 1
            self.waiting = False
            return True
        if op == "add":
            self.regs[x] = self.regs.get(x, 0) + val(y or "0", self.regs)
            self.ip += 1
            self.waiting = False
            return True
        if op == "mul":
            self.regs[x] = self.regs.get(x, 0) * val(y or "0", self.regs)
            self.ip += 1
            self.waiting = False
            return True
        if op == "mod":
            self.regs[x] = self.regs.get(x, 0) % val(y or "1", self.regs)
            self.ip += 1
            self.waiting = False
            return True
        if op == "rcv":
            if self.inq:
                self.regs[x] = self.inq.popleft()
                self.ip += 1
                self.waiting = False
                return True
            self.waiting = True
            return False
        if op == "jgz":
            if val(x, self.regs) > 0:
                self.ip += val(y or "0", self.regs)
            else:
                self.ip += 1
            self.waiting = False
            return True

        raise ValueError("bad opcode")


def solve_part2(prog: list[tuple[str, str, str | None]]) -> int:
    p0 = Proc(0, prog)
    p1 = Proc(1, prog)

    while True:
        progressed0 = p0.step(p1.inq)
        progressed1 = p1.step(p0.inq)

        if (not progressed0 and not progressed1) and (p0.waiting or p0.terminated) and (p1.waiting or p1.terminated):
            return p1.send_count


def main() -> int:
    ap = argparse.ArgumentParser(description="AoC 2017 Day 18 fancy Python")
    ap.add_argument("--part", type=int, required=True, choices=[1, 2])
    ap.add_argument("--input")
    args = ap.parse_args()

    input_path = Path(args.input).resolve() if args.input else default_input_path()
    got_sha = sha256_file(input_path)
    if got_sha != EXPECTED_SHA256:
        raise SystemExit(f"checksum mismatch: {got_sha} != {EXPECTED_SHA256}")

    prog = parse(input_path.read_text(encoding="utf-8"))

    t0 = time.perf_counter_ns()
    answer = str(solve_part1(prog) if args.part == 1 else solve_part2(prog))
    runtime_ms = (time.perf_counter_ns() - t0) / 1e6

    expected = EXPECTED_PART1 if args.part == 1 else EXPECTED_PART2
    if answer != expected:
        raise SystemExit(f"answer mismatch for part {args.part}: got {answer!r}, expected {expected!r}")

    print(answer)
    print(f"[python-fancy] day=18 part={args.part} runtime_ms={runtime_ms:.3f}", file=__import__("sys").stderr)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: advent2017/extras/rust/src/bin/day18.rs
// Architecture Notes:
// - This file is heavily commented for long-term maintainability and reconstruction.
// - Pipeline shape is parse -> model -> compute -> emit.
// - Performance-sensitive sections document data-layout and concurrency tradeoffs.

use sha2::{Digest, Sha256};
use std::collections::VecDeque;
use std::env;
use std::fs;
use std::path::Path;
use std::time::Instant;

const EXPECTED_SHA: &str = "4052bdd33baaf7be897365aa3ad1cff5fae76ade4c474c9e5ebcdf5058ad368e";
const EXPECTED_PART1: &str = "7071";
const EXPECTED_PART2: &str = "8001";

#[derive(Clone)]
struct Inst {
    op: String,
    x: String,
    y: Option<String>,
}

fn resolve_input(provided: &str) -> String {
    if !provided.is_empty() { return provided.to_string(); }
    let cands = ["advent2017/Day18/d18_input.txt", "Day18/d18_input.txt", "../Day18/d18_input.txt", "../../Day18/d18_input.txt"];
    for c in cands { if Path::new(c).exists() { return c.to_string(); } }
    panic!("input not found");
}

fn sha256_file(path: &str) -> String {
    let b = fs::read(path).expect("read");
    let mut h = Sha256::new(); h.update(&b); hex::encode(h.finalize())
}

fn parse(raw: &str) -> Vec<Inst> {
    raw.lines()
        .filter_map(|line| {
            let line = line.trim();
            if line.is_empty() { None }
            else {
                let p = line.split_whitespace().collect::<Vec<&str>>();
                Some(Inst {
                    op: p[0].to_string(),
                    x: p[1].to_string(),
                    y: if p.len() > 2 { Some(p[2].to_string()) } else { None },
                })
            }
        })
        .collect::<Vec<Inst>>()
}

fn val(tok: &str, regs: &std::collections::HashMap<String, i64>) -> i64 {
    if let Ok(v) = tok.parse::<i64>() { v } else { *regs.get(tok).unwrap_or(&0) }
}

fn solve_part1(prog: &[Inst]) -> i64 {
    let mut regs: std::collections::HashMap<String, i64> = std::collections::HashMap::new();
    let mut ip: i64 = 0;
    let mut last: i64 = 0;

    while ip >= 0 && (ip as usize) < prog.len() {
        let inx = &prog[ip as usize];
        match inx.op.as_str() {
            "snd" => last = val(&inx.x, &regs),
            "set" => { regs.insert(inx.x.clone(), val(inx.y.as_ref().unwrap(), &regs)); }
            "add" => {
                let cur = *regs.get(&inx.x).unwrap_or(&0);
                regs.insert(inx.x.clone(), cur + val(inx.y.as_ref().unwrap(), &regs));
            }
            "mul" => {
                let cur = *regs.get(&inx.x).unwrap_or(&0);
                regs.insert(inx.x.clone(), cur * val(inx.y.as_ref().unwrap(), &regs));
            }
            "mod" => {
                let cur = *regs.get(&inx.x).unwrap_or(&0);
                regs.insert(inx.x.clone(), cur % val(inx.y.as_ref().unwrap(), &regs));
            }
            "rcv" => { if val(&inx.x, &regs) != 0 { return last; } }
            "jgz" => {
                if val(&inx.x, &regs) > 0 {
                    ip += val(inx.y.as_ref().unwrap(), &regs);
                    continue;
                }
            }
            _ => panic!("bad op"),
        }
        ip += 1;
    }

    panic!("no recovery");
}

struct Proc {
    regs: std::collections::HashMap<String, i64>,
    ip: i64,
    inq: VecDeque<i64>,
    send_count: i64,
    waiting: bool,
    terminated: bool,
}

fn step(me: &mut Proc, other: &mut Proc, prog: &[Inst]) -> bool {
    if me.ip < 0 || (me.ip as usize) >= prog.len() {
        me.terminated = true;
        me.waiting = false;
        return false;
    }

    let inx = &prog[me.ip as usize];
    match inx.op.as_str() {
        "snd" => {
            other.inq.push_back(val(&inx.x, &me.regs));
            me.send_count += 1;
            me.ip += 1;
            me.waiting = false;
            true
        }
        "set" => {
            me.regs.insert(inx.x.clone(), val(inx.y.as_ref().unwrap(), &me.regs));
            me.ip += 1;
            me.waiting = false;
            true
        }
        "add" => {
            let cur = *me.regs.get(&inx.x).unwrap_or(&0);
            me.regs.insert(inx.x.clone(), cur + val(inx.y.as_ref().unwrap(), &me.regs));
            me.ip += 1;
            me.waiting = false;
            true
        }
        "mul" => {
            let cur = *me.regs.get(&inx.x).unwrap_or(&0);
            me.regs.insert(inx.x.clone(), cur * val(inx.y.as_ref().unwrap(), &me.regs));
            me.ip += 1;
            me.waiting = false;
            true
        }
        "mod" => {
            let cur = *me.regs.get(&inx.x).unwrap_or(&0);
            me.regs.insert(inx.x.clone(), cur % val(inx.y.as_ref().unwrap(), &me.regs));
            me.ip += 1;
            me.waiting = false;
            true
        }
        "rcv" => {
            if let Some(v) = me.inq.pop_front() {
                me.regs.insert(inx.x.clone(), v);
                me.ip += 1;
                me.waiting = false;
                true
            } else {
                me.waiting = true;
                false
            }
        }
        "jgz" => {
            if val(&inx.x, &me.regs) > 0 {
                me.ip += val(inx.y.as_ref().unwrap(), &me.regs);
            } else {
                me.ip += 1;
            }
            me.waiting = false;
            true
        }
        _ => panic!("bad op"),
    }
}

fn solve_part2(prog: &[Inst]) -> i64 {
    let mut p0 = Proc {
        regs: std::collections::HashMap::from([("p".to_string(), 0)]),
        ip: 0,
        inq: VecDeque::new(),
        send_count: 0,
        waiting: false,
        terminated: false,
    };
    let mut p1 = Proc {
        regs: std::collections::HashMap::from([("p".to_string(), 1)]),
        ip: 0,
        inq: VecDeque::new(),
        send_count: 0,
        waiting: false,
        terminated: false,
    };

    loop {
        let a = step(&mut p0, &mut p1, prog);
        let b = step(&mut p1, &mut p0, prog);
        if !a && !b && (p0.waiting || p0.terminated) && (p1.waiting || p1.terminated) {
            return p1.send_count;
        }
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let mut part = 0;
    let mut input = String::new();
    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--part" => { part = args[i + 1].parse().unwrap(); i += 2; }
            "--input" => { input = args[i + 1].clone(); i += 2; }
            _ => panic!("unknown arg"),
        }
    }
    if part != 1 && part != 2 { panic!("--part 1|2 required"); }

    let in_path = resolve_input(&input);
    if sha256_file(&in_path) != EXPECTED_SHA { panic!("checksum mismatch"); }
    let prog = parse(&fs::read_to_string(&in_path).expect("read_to_string"));

    let t0 = Instant::now();
    let ans = if part == 1 { solve_part1(&prog).to_string() } else { solve_part2(&prog).to_string() };
    let expected = if part == 1 { EXPECTED_PART1 } else { EXPECTED_PART2 };
    if ans != expected { panic!("answer mismatch"); }

    println!("{}", ans);
    let ms = t0.elapsed().as_nanos() as f64 / 1e6;
    eprintln!("[rust-fancy] day=18 part={} runtime_ms={:.3}", part, ms);
}
